### DISCUSSION TITLE
Raspberry Pi + Docker How-to (for HASS users)

### DISCUSSION THREAD

**Discussion by scyto** [zwave-js/zwave-js-ui#180] [Show and tell]
# Introduction
This how gives steps to getting docker and zwavejs2mqtt working on a raspberry pi 4.

Assumptions in this example:
1. I am using a RazBerry zwave pi-hat - this will use /dev/ttyAMA0, if using a USB change serial port as needed in files and commands
2. This is a dedicated z-wave node not running anything else
3. I want the UI on port 80 so i don't have to put port numbers in my browser (i am lazy)
4. you want to pass through the devices to a working HASS instance

# Prepare Raspberry Pi
## Basic setup
Fit raZberry hat to pi, then plug in power
Image Raspbian image and boot and logon (if you don't know how to do this, go find a pi and linux tutorial)
login

	sudo apt update
	sudo apt upgrade

## Give your pi a friendly name
use raspi-config to set name to something like pi-zwave01

 	choose option 1 system options
 	choose option S4 Hostname
don't forget to save as you backout of the menus
> Note: latest raspbian has avahai loaded by default which has mDNS enabled; meaning you can address your pi as http://pi-zwave01 rather than using IP address if your client is mDNS aware without needing to make a DNS entry in your router/dns server.  

## Disable Bluetooth on pi4
 only needed if using pihat, see here for detailed why https://di-marco.net/blog/it/2020-06-06-paspberry_pi_3_4_and_0_w_serial_port_usage/
 
 	sudo nano /boot/config.txt
 
 add following and save and exit
 
	# Disable Bluetooth
	dtoverlay=disable-bt
	
# Install docker
	cd ~
 	curl -fsSL https://get.docker.com -o get-docker.sh
 	sudo sh get-docker.sh
 	sudo usermod -aG docker pi
 	sudo apt install python3 
 	sudo apt install python3.pip
 	sudo pip3 -v install docker-compose  (is the -v really needed?)

# Configure zwavejs2mqtt container

	cd ~
 	mkdir zwavejs2mqtt
	cd zwavejs2mqtt
	curl -fsSL https://raw.githubusercontent.com/zwave-js/zwavejs2mqtt/master/docker/docker-compose.yml -o docker-compose.yml
	nano docker-compose.yml
		○ Change serial in devices to correct serial port (e.g. /dev/ttyAMA0)
		○ Change port to 80:8091 (optional, but rest of example assumes this change was made)
		○ Exit nano and save
	Sudo docker-compose up -d
	
# Configure zwavejs2mqtt
Connect to http://pi-zwave01
1. click the cog 
2. click zwave

The following fields need to be filled:

	• Serial port with your serial port path inside the container - e.g /dev/ttyAMA0
	• Poll interval 1000ms (aka 1s)
	• Commands timeout 20 seconds
	• click the symbol to the right of network key to generate a secure network key (or enter your own)
  • click save

If this is done right after a few mins when you return to the main page the HomeID and Home Hex should be populated, if they are not refer to troubleshooting guide.

# Configure MQTT
Assumes you are already running an MQTT container like the one in home assistant supervisor mode or mosquitto from docker

Click the cog in the UI
expand mqtt
Set as follows:

	Name = anything you want, this is just an identifier
	host url = mqt://<ip address of MQTT instance>
	port = - 1883 (this is default for mosquitto etc, change as needed)
	reconnect period (ms) = 500
	prefix = zwavejs2mqtt (this can be anything you want and must not conflict)
	qos = 1 (i saw somewhere a recommendation it should be 1, not sure if thats true)
	retain = enabled (as recommended for Hass)
	store = enabled (i have no idea why)
	clean = disabled (i have no idea why)
	auth = disabled (only needed if you configured your MQTT to use auth, by default things like mosquitto don't seem to be configured that way)
Click Save!

# Configure Gateway (assuming for Home Assistant)
Click the cog in the UI
expand Gateway
Set as follows:

	Type = ValueID topics
	Payload type = Entire Z-Wave Value Object
	Use node names instead of numeric node ID = disabled (i prefer it enabled, but there are issues updating the nodes names, so leave ad disabled unless you really want this)
	Send zwave events = disabled (i don't why)
	Include Node info = enabled
	discovery prefix = zwavejs2mqtt (no idea why, copied from an article i saw)
	ignore location = disabled
	ignore status updated = disabled 
	Hass discovery = enabled (because this guide assume Hass)
	Retained discovery = disabled (someonw told me to set this this way, not sure what best practice is)
Click Save!

Don't forget to make sure you setup MQTT in Hass configuration.yaml as per instructions https://zwave-js.github.io/zwavejs2mqtt/#/guide/homeassistant

let me know what you think the best settings are and why!

**Comment by scyto** [comment]
and as this discussion is classified *show* and tell  (this is from top layer to bottom layer 1)zwave module 2)POE+i2c+screen module 3) pi4 2gb 
![image](https://user-images.githubusercontent.com/11369180/104073909-bd387b00-51c3-11eb-8f81-95ed8f0e744a.png)

**Comment by LordMike** [comment]
How is the range and stability of it?

I considered investing in raspberry to run z-wave alone, and possibly Home Assistant as well, and I found that hat. My current situation doesn't allow for the z-wave controller to be placed centrally in the house, which sucks.

I didn't know if it would be good or not.. I'd be oming from the Aeotec gen5 stick, which has a built in battery... Which in turn does _not_ work on the raspberry pi 4, due to some hardware issue.. A newer model would apparently fix that.

Mike.

**Comment by stefanoklett** [comment]
by me it doesn't work, what could it be?
```
pi@razberry:~/zwavejs2mqtt $ sudo docker-compose up -d
/usr/lib/python3/dist-packages/requests/__init__.py:91: RequestsDependencyWarning: urllib3 (2.0.0) or chardet (3.0.4) doesn't match a supported version!
  RequestsDependencyWarning)
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/docker/api/client.py", line 214, in _retrieve_server_version
    return self.version(api_version=False)["ApiVersion"]
  File "/usr/local/lib/python3.7/dist-packages/docker/api/daemon.py", line 181, in version
    return self._result(self._get(url), json=True)
  File "/usr/local/lib/python3.7/dist-packages/docker/utils/decorators.py", line 46, in inner
    return f(self, *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/docker/api/client.py", line 237, in _get
    return self.get(url, **self._set_request_timeout(kwargs))
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/usr/local/lib/python3.7/dist-packages/urllib3/connectionpool.py", line 802, in urlopen
    **response_kw,
  File "/usr/local/lib/python3.7/dist-packages/urllib3/connectionpool.py", line 504, in _make_request
    enforce_content_length=enforce_content_length,
TypeError: request() got an unexpected keyword argument 'chunked'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/docker-compose", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/dist-packages/compose/cli/main.py", line 81, in main
    command_func()
  File "/usr/local/lib/python3.7/dist-packages/compose/cli/main.py", line 200, in perform_command
    project = project_from_options('.', options)
  File "/usr/local/lib/python3.7/dist-packages/compose/cli/command.py", line 70, in project_from_options
    enabled_profiles=get_profiles_from_options(options, environment)
  File "/usr/local/lib/python3.7/dist-packages/compose/cli/command.py", line 153, in get_project
    verbose=verbose, version=api_version, context=context, environment=environment
  File "/usr/local/lib/python3.7/dist-packages/compose/cli/docker_client.py", line 43, in get_client
    environment=environment, tls_version=get_tls_version(environment)
  File "/usr/local/lib/python3.7/dist-packages/compose/cli/docker_client.py", line 170, in docker_client
    client = APIClient(use_ssh_client=not use_paramiko_ssh, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/docker/api/client.py", line 197, in __init__
    self._version = self._retrieve_server_version()
  File "/usr/local/lib/python3.7/dist-packages/docker/api/client.py", line 222, in _retrieve_server_version
    f'Error while fetching server API version: {e}'
docker.errors.DockerException: Error while fetching server API version: request() got an unexpected keyword argument 'chunked'
```

regards
Stefano

**Reply by LordMike** [reply]
That seems to be a docker issue. You’ll probably find that all docker compose commands are failing.

**Reply by stefanoklett** [reply]
I rebuilt my raspberry PI from the scratch, now it seems to be working. Before somehow it was an old version that probably didn't have everything needed on board

regards
